### ISSUE TITLE
The "data" member of PaymentMethodData doesn't seem to actually be used for anything

### ISSUE THREAD

**Issue by bzbarsky** [w3c/browser-payment-api#338]
The PaymentRequest constructor enforces it being "JSON-serializable", whatever that means. .

But after that, nothing references that member.  Is it supposed to actually get serialized to JSON at some point, perhaps?

**Comment by adamroach** [comment]
The information that goes in `data` is specific to individual payment methods. See, for example, the [relevant section in the basic card specification](https://w3c.github.io/webpayments-methods-card/#request).

The requirement to be serializable is to allow payment apps to send them to a server for subsequent processing (if necessary) and to allow sending them between processes in some compatible way for communication with native OS wallets (e.g., for invoking Android native payment apps).

**Comment by bzbarsky** [comment]
> The information that goes in data is specific to individual payment methods.

That's fine, but nothing in this spec says the "data" is even passed to any other specification's algorithm.

And since examining the data can have script-observable side-effects, you have to do it at some well-defined point.   And probably before going async, yes?

**Comment by domenic** [comment]
> See, for example, the relevant section in the basic card specification.

This link appears broken. Assuming it is meant to go to https://w3c.github.io/webpayments-methods-card/#basiccardrequest, @bzbarsky's point still stands. Nothing in the spec actually uses that dictionary. In https://github.com/w3c/browser-payment-api/pull/381 we're adding validation that the data is in the correct format, but we're still never using the data. So as far as I can tell it is conformant to implement the specification by completely ignoring the data member, e.g. by accepting debit cards when the site specifies it only accepts credit cards.

To fix this, probably what is necessary is to insert a step into https://w3c.github.io/browser-payment-api/#user-accepts-the-payment-request-algorithm which passes data to some algorithm defined in each payment method specification, and then the payment method specification for basic-card can say something like "use supportedMethods and supportedTypes to determine an appropriate response by prompting the user." That algorithm will also probably be what creates the appropriate payment response details vaguely alluded to in step 7:

> Set the details attribute value of response to a JSON-serializable object containing the payment method specific message that will be used by the merchant to process the transaction. The format of this response will be defined for each payment method. 

(Currently basic card definitely does not define a format for that response. It defines a BasicCardResponse dictionary, but never says that an instance of that dictionary should be created and used in the algorithm, and never says how to populate the values.)

**Comment by domenic** [comment]
Also, it might be better to just do the dictionary conversion once, e.g. in the constructor, instead of doing it once in the constructor for validation, and then a second time as part of the "user accepts the payment request" algorithm. This is observably different; I wonder what Blink and Edge do.

**Comment by dlongley** [comment]
> To fix this, probably what is necessary is to insert a step into https://w3c.github.io/browser-payment-api/#user-accepts-the-payment-request-algorithm which passes data to some algorithm defined in each payment method specification...

Or passes it to some algorithm defined in the Payment App spec. My understanding is that that's where the data is actually going -- to a selected Payment App. Only in the case of a Payment App that is "built-in" to the user agent will it do something different (but this difference could be abstracted away).

**Comment by domenic** [comment]
Yeah, maybe; I'm new to all this and don't quite understand the distinction. I've only ever seen the basic-card spec and how I thought that integrates into the flow. Is there some kind of payment app spec for card processing that I should be looking at instead?

**Comment by rsolomakhin** [comment]
I'm happy to provide more background information. :-)

Although Chrome's implementation of "basic-card" is based on Autofill data in Chrome, this is done via a "payment app" that we call "Autofill Payment App." It should behave as any 3rd party payment app that's being specified in [Payment App API](https://w3c.github.io/webpayments-payment-apps-api/) spec. Once that spec is finished and implemented, any 3rd party payment app should be able to register to handle "basic-card".

Note that Android Pay integration into Chrome is also done via a "payment app" that we call "Android Pay Payment App." This is temporary until the [Android Payment Apps](https://groups.google.com/a/chromium.org/forum/#!msg/chromium-dev/fsslHD1Gf88/K2KpikS6BwAJ) support is implemented. Android Pay implementation does not have a W3C-style spec, but they do have a [tutorial](https://developers.google.com/web/fundamentals/discovery-and-monetization/payment-request/android-pay#integration_using_gateway_token).

By the way, someone from Samsung has written up a [SamsungPay Payment Method](https://maheshkk.github.io/webpayments/proposals/SamsungPay-payment-method.html) spec that can be useful to read and compare to "basic-card".

**Comment by dlongley** [comment]
@domenic,

You may find it helpful to look at the Web Payments Overview doc as well, in particular, the roles in the ecosystem:

https://www.w3.org/TR/webpayments-overview/#roles-in-the-ecosystem

**Comment by dlongley** [comment]
The user agent primarily plays the role of the "Mediator".

**Comment by domenic** [comment]
Hmm, OK, that makes sense! So I think the flow in terms of algorithms is that this spec should call into the payment method specs, which can then delegate to whatever payment app is currently registered to handle that payment method.

And I guess that, per the discussions in https://github.com/w3c/browser-payment-api/issues/307#issuecomment-267145043, the payment method spec is responsible for some preprocessing before it hands off the data to the payment app. E.g. basic-card will parse the data string into a BasicCardRequest dictionary.

**Comment by dlongley** [comment]
> Hmm, OK, that makes sense! So I think the flow in terms of algorithms is that this spec should call into the payment method specs, which can then delegate to whatever payment app is currently registered to handle that payment method.

Hmm, I don't think that quite fits with the architecture. The flow is like this, at a high level:

1. The merchant website invokes `show()` on a PaymentRequest and gets a Promise.
2. The user agent shows the user all of the Payment Apps that could be used to fulfill the PaymentRequest.
3. The user selects the Payment App they want to pay with.
4. The user agent sends the serialized data to the Payment App. Note: How the data gets to the Payment App depends on its type. It may be a Web-based Payment App, which is a ServiceWorker and the data will be passed through an event to it. It may be a native app or some built-in app (or something else proprietary) and the user agent will figure out how to get the data to it on its own.
5. The Payment App, which implements the payment method (based on its particular spec), does whatever it needs to do to produce a PaymentResponse, which is sent back through the Mediator (user agent) to resolve the Promise.

**Comment by dlongley** [comment]
Note that this architecture is meant to support many different payment methods -- and the user agent may only ever support the simple "basic card" (or none at all). It is up to Payment Apps to implement payment methods. These payment methods may not use any Web IDL or do anything remotely browser related.

**Comment by domenic** [comment]
> The user agent shows the user all of the Payment Apps that could be used to fulfill the PaymentRequest.

This is the key step that is missing some detail. From what I understand from https://github.com/w3c/browser-payment-api/issues/307#issuecomment-267156994 , finding out what Payment Apps can be used to fulfill that payment request requires asking the payment method to do some preprocessing first. E.g. for basic card, trying to convert data (which at this point is a string) to a BasicCardRequest. Only if this succeeds will basic card then ask the payment app if it can handle the payment request. In the payment app spec that would be by calling canHandle(supportedMethods, shoppingCart) where supportedMethods contains a PaymentMethodData containing the parsed BasicCardRequest.

**Comment by rsolomakhin** [comment]
> Finding out what Payment Apps can be used to fulfill that payment request requires asking the payment method to do some preprocessing first.

@domenic, I think payment methods specs are implemented by payment apps. In a sense, a payment method is an interface and the payment apps are implementations. So "asking the payment method to do some preprocessing" is equivalent to saying "asking the payment apps to do some processing." The way we achieve this is through JSON-serializing the method-specific data, passing that over IPC to the payment apps, which then JSON-parse the data for their own needs.

**Comment by domenic** [comment]
Hmm, OK. I guess that makes these things a little tricky to specify then. E.g. if it's the payment app that does the processing into a BasicCardRequest, and the payment app is a UA-specific thing (e.g. Chrome's "Autofill basic card payment app" you mentioned), then there's no specification that ever actually uses BasicCardRequest :(.

**Comment by rsolomakhin** [comment]
Can we say something like this in "basic-card" spec?

> The payment app that supports "basic-card" expects to be invoked with `supportedMethods.data` that can be parsed into `BasicCardRequest`.

It's a bit of a mind-bender. Sorry! :-(

**Comment by domenic** [comment]
Yeah, I'll try to give it a shot!

**Comment by delapuente** [comment]
> To fix this, probably what is necessary is to insert a step into https://w3c.github.io/browser-payment-api/#user-accepts-the-payment-request-algorithm which passes data to some algorithm defined in each payment method specification

This should be ideal and providing an example would be also an interesting addition. It took a bit for me, until I read the [integration guide with Android Pay](https://developers.google.com/web/fundamentals/discovery-and-monetization/payment-request/android-pay), to realise that `data` served a purpose. 

Another issue related to this is the fact that the `data` member affects all the payment method ids in the `supportedMethods` array. What is the use case for this? I would rather expect a pair `{ id, data } for each _paymentMethod_.

**Comment by rsolomakhin** [comment]
> Another issue related to this is the fact that the data member affects all the payment method ids in the supportedMethods array. What is the use case for this?

Some payment methods might have identical parameters in `data`, so sharing is useful there.

**Comment by delapuente** [comment]
I imagine that. Thank you for the clarification. Even so, we could provide an example.

**Comment by rsolomakhin** [comment]
Here's an example for a merchant that accepts only debit cards, regardless of whether they are passed in the raw via `'basic-card'` or encrypted via `'tokenized-card'`.

```javascript
new PaymentRequest(
    [{
        supportedMethods: ['basic-card', 'tokenized-card'],
        data: {supportedTypes: ['debit']}
    }],
    shoppingCart);
```

**Comment by delapuente** [comment]
Really appreciated. Do you think it would be a good idea to add the example
to the spec?

**Comment by rsolomakhin** [comment]
Could not hurt.
